drivers/at_cmd/at_cmd.c

```diff
@@ -51,6 +59,94 @@ struct callback_work_item {
 K_MEM_SLAB_DEFINE(rsp_work_items, sizeof(struct callback_work_item),
 		  CONFIG_AT_CMD_RESPONSE_BUFFER_COUNT, 4);
 
+/**@brief Find notification handler.
+ *
+ * @note  Returns next and writes previous to prev.
+ */
+static struct notif_handler *find_notif_handler(sys_slist_t *list,
+	at_cmd_handler_t handler, struct notif_handler **prev_out)
+{
+	struct notif_handler *prev = NULL;
+	struct notif_handler *curr, *tmp;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(list, curr, tmp, node) {
+		if (curr->handler == handler) {
+			*prev_out = prev;
+			return curr;
+		}
+		prev = curr;
+	}
+	return NULL;
+}
+
+/**@brief Append notification handler. */
+static int append_notif_handler(sys_slist_t *list, at_cmd_handler_t handler)
+{
+	struct notif_handler *to_ins;
+
+	k_mutex_lock(&list_mtx, K_FOREVER);
+
+	/* Check if it exists */
+	if (find_notif_handler(list, handler, &to_ins) != NULL) {
+		LOG_DBG("Already in the list.\n");
+		k_mutex_unlock(&list_mtx);
+		return 0;
+	}
+
+	/* Allocate memory and fill. */
+	to_ins = (struct notif_handler *)k_malloc(sizeof(struct notif_handler));
```

Comment:
@rlubos Doesn't this going to give us the same headache as before?

Reply:
`struct notif_handler` is a small structure - 8 bytes, previously we've dealt with over 2,5K byte buffers. It will not hog the heap even with multiple allocations. Perhaps a preallocated mem slab for handlers would be a better approach, I don't find it to be a blocker.